enterprise_access/apps/api/serializers/subsidy_access_policy.py

```diff
@@ -87,13 +107,28 @@ def create(self, validated_data):
 
     def validate(self, attrs):
         super().validate(attrs)
-        policy_type = attrs.get('policy_type', None)
-        if policy_type:
-            # just some extra caution around discarding the other two credit limits if they have a non-zero value
-            if policy_type in POLICY_TYPES_WITH_CREDIT_LIMIT:
-                for field in POLICY_TYPE_CREDIT_LIMIT_FIELDS:
-                    if field != POLICY_TYPE_FIELD_MAPPER.get(policy_type):
-                        attrs.pop(field)
+
+        # Get the policy subclass.
+        # super().validate() already checked that attrs contains a "policy_type" key and valid value.
+        policy_type = attrs['policy_type']
+        policy_class = SubsidyAccessPolicy.get_policy_class_by_type(policy_type)
+
+        # Must specify exactly the required custom fields as declared by the policy subclass, no more no less.
+        custom_policy_field_errors = []
+        # Here's the "no less" part:
+        if not set(policy_class.REQUIRED_CUSTOM_FIELDS).issubset(attrs.keys()):
+            custom_policy_field_errors.append(
+                f"Missing fields for {policy_type} policy type: {policy_class.REQUIRED_CUSTOM_FIELDS}."
+            )
+        # Here's the "no more" part:
+        unused_custom_fields = set(policy_class.ALL_CUSTOM_FIELDS) - set(policy_class.REQUIRED_CUSTOM_FIELDS)
+        if unused_custom_fields.intersection(attrs.keys()):
+            custom_policy_field_errors.append(
+                f"Extraneous fields for {policy_type} policy type: {list(unused_custom_fields)}."
+            )
+        if custom_policy_field_errors:
+            raise serializers.ValidationError(custom_policy_field_errors)
```

Comment:
Would it be simpler to do something like
```python
if set(attrs.keys()) != set(policy_class.REQUIRED_CUSTOM_FIELDS):
    errors.append(f"Provided fields {attrs.keys()} does not exactly match the required custom fields {policy_class.REQUIRED_CUSTOM_FIELDS}")
```

Reply:
attrs.keys() likely contains a bunch of other non-custom fields such as "enterprise_customer_uuid" and "catalog_uuid" though.   That said, if I consolidated this code to just one error it would be a shorter function.